pkg/commands/kvstoreentry/create.go

```diff
@@ -26,20 +28,233 @@ func NewCreateCommand(parent cmd.Registerer, g *global.Data, m manifest.Data) *C
+		return fsterr.ErrInvalidKVCombo
+	}
+
 	err := c.Globals.APIClient.InsertKVStoreKey(&c.Input)
 	if err != nil {
 		c.Globals.ErrLog.Add(err)
 		return err
 	}
 
-	text.Success(out, "Inserted key %s into kv store %s", c.Input.Key, c.Input.ID)
+	text.Success(out, "Inserted key %s into KV Store %s", c.Input.Key, c.Input.ID)
+
+	return nil
+}
+
+// CheckFlags ensures only one of the three specified flags are provided.
+func (c *CreateCommand) CheckFlags() error {
+	flagCount := 0
+	if c.stdin {
+		flagCount++
+	}
+	if c.filePath != "" {
+		flagCount++
+	}
+	if c.dirPath != "" {
+		flagCount++
+	}
+	if flagCount > 1 {
+		return fsterr.ErrInvalidStdinFileDirCombo
+	}
 	return nil
 }
+
+// ProcessStdin streams STDIN to the batch API endpoint.
+func (c *CreateCommand) ProcessStdin(in io.Reader, out io.Writer) error {
+	// Determine if 'in' has data available.
+	if in == nil || text.IsTTY(in) {
+		return fsterr.ErrNoSTDINData
+	}
+	return c.CallBatchEndpoint(in, out)
+}
+
+// ProcessFile streams a JSON file content to the batch API endpoint.
+func (c *CreateCommand) ProcessFile(out io.Writer) error {
+	f, err := os.Open(c.filePath)
+	if err != nil {
+		c.Globals.ErrLog.Add(err)
+		return err
+	}
+	return c.CallBatchEndpoint(f, out)
+}
+
+// ProcessDir concurrently reads files from the given directory structure and
+// uploads each file to the set-value-for-key endpoint where the filename is the
+// key and the file content is the value.
+//
+// NOTE: Unlike ProcessStdin/ProcessFile content doesn't need to be base64.
+func (c *CreateCommand) ProcessDir(out io.Writer) error {
+	path, err := filepath.Abs(c.dirPath)
+	if err != nil {
+		return err
+	}
+
+	files, err := os.ReadDir(path)
+	if err != nil {
+		return err
+	}
+
+	spinner, err := text.NewSpinner(out)
+	if err != nil {
+		return err
+	}
+
+	err = spinner.Start()
+	if err != nil {
+		return err
+	}
+	fileLength := len(files)
+	msg := "Processed %d of %d files"
+	spinner.Message(fmt.Sprintf(msg, 0, fileLength) + "...")
+
+	base := filepath.Base(path)
+	errors := []ProcessErr{}
+	processed := make(chan struct{}, c.dirConcurrency)
+	sem := make(chan struct{}, c.dirConcurrency)
+
+	var (
+		filesProcessed uint64
+		wg             sync.WaitGroup
+	)
+
+	go func() {
+		for range processed {
+			filesProcessed++
+			spinner.Message(fmt.Sprintf(msg, filesProcessed, fileLength) + "...")
+		}
+	}()
+
+	for _, file := range files {
+		if !file.IsDir() {
+			wg.Add(1)
+
+			go func(file fs.DirEntry) {
+				// Restrict resource allocation if concurrency limit is exceeded.
+				sem <- struct{}{}
+				defer func() {
+					<-sem
+				}()
+				defer wg.Done()
+
+				filePath := filepath.Join(c.dirPath, file.Name())
+				dir, filename := filepath.Split(filePath)
+				index := strings.Index(dir, base)
+				filename = filepath.Join(dir[index:], filename)
+
+				// G304 (CWE-22): Potential file inclusion via variable
+				// #nosec
+				fileContent, err := os.ReadFile(filePath)
```

Comment:
I'm guessing that these files could be quite large, and with the default concurrency of 100, could result in a lot of memory usage reading the entire contents into memory.

I'm not though if there's an alternate approach that's possible. Would require a kv store API that accepts something like an `io.Reader`

Reply:
There is a batch API endpoint (not documented) but it was problematic due to the need to be able to adequately audit what specific files failed and with the batch endpoint it wasn't possible to determine that without making additional API calls to list all keys after each batch call which itself is timely/costly. So we opted to avoid the batch endpoint for concurrent processing.

I'll have a think about this more.

Reply:
I have a draft PR for go-fastly trying to figure out what's possible:
https://github.com/fastly/go-fastly/pull/437

Reply:
@awilliams-fastly I've got the changes to use the go-fastly PR locally and I'll push them up once you've had a chance to review/approve the go-fastly PR. Thanks!